timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/AsyncTimelockServiceTargetedSweepRateLimitingTest.java

```diff
@@ -0,0 +1,177 @@
+        Pair<Meter, ListenableFuture<?>> nonTsLock2 = run(() -> nonTsLockRequest(), CLIENT, executorService);
+        Pair<Meter, ListenableFuture<?>> tsLock =
+                run(ShardAndStrategy.of(0, CONSERVATIVE), CLIENT, executorService);
+        Pair<Meter, ListenableFuture<?>> tsLockDuplicated =
+                run(ShardAndStrategy.of(0, CONSERVATIVE), CLIENT, executorService);
+        Pair<Meter, ListenableFuture<?>> tsLock2 =
+                run(ShardAndStrategy.of(0, THOROUGH), CLIENT, executorService);
+
+        Pair<Meter, ListenableFuture<?>> nonTsLock_r =
+                run(() -> nonTsLockRequest(), RATE_LIMITED_CLIENT, executorService2);
+        Pair<Meter, ListenableFuture<?>> nonTsLock2_r =
+                run(() -> nonTsLockRequest(), RATE_LIMITED_CLIENT, executorService2);
+        Pair<Meter, ListenableFuture<?>> tsLock_r =
+                run(ShardAndStrategy.of(0, CONSERVATIVE), RATE_LIMITED_CLIENT, executorService2);
+        Pair<Meter, ListenableFuture<?>> tsLockDuplicated_r =
+                run(ShardAndStrategy.of(0, CONSERVATIVE), RATE_LIMITED_CLIENT, executorService2);
+        Pair<Meter, ListenableFuture<?>> tsLock2_r =
+                run(ShardAndStrategy.of(0, THOROUGH), RATE_LIMITED_CLIENT, executorService2);
+
+        List<Pair<Meter, ListenableFuture<?>>> nonRateLimitedRuns =
+                ImmutableList.of(nonTsLock, nonTsLock2, tsLock, tsLockDuplicated, tsLock2);
+
+        List<Pair<Meter, ListenableFuture<?>>> rateLimitedRuns =
+                ImmutableList.of(nonTsLock_r, nonTsLock2_r, tsLock_r, tsLockDuplicated_r, tsLock2_r);
+
+        Future<List<Object>> combinedFuture = asFuture(Iterables.concat(nonRateLimitedRuns, rateLimitedRuns));
+
+        Thread.sleep(15_000);
+        combinedFuture.cancel(true);
+
+        SoftAssertions softAssertions = new SoftAssertions();
+        softAssertions.assertThat(meanRates(nonRateLimitedRuns))
+                .as("None of these should be rate limited")
+                .allSatisfy(AsyncTimelockServiceTargetedSweepRateLimitingTest::isNotRateLimited);
+
+        List<Double> rateLimitedRunMeanRates = meanRates(rateLimitedRuns);
+        softAssertions.assertThat(rateLimitedRunMeanRates.subList(0, 2))
+                .as("first two non ts lock requests")
+                .allSatisfy(AsyncTimelockServiceTargetedSweepRateLimitingTest::isNotRateLimited);
+
+        softAssertions.assertThat(rateLimitedRunMeanRates.get(4))
+                .as("last ts non-shared request is in rate limited correctly")
+                .isLessThan(LOCK_ACQUIRES_PER_SECOND + Offset.offset(0.3).value);
+
+        softAssertions.assertThat(rateLimitedRunMeanRates.get(2) + rateLimitedRunMeanRates.get(3))
+                .as("adding rates for two separate requests for same lock should be within rate limit")
+                .isLessThan(LOCK_ACQUIRES_PER_SECOND + Offset.offset(0.3).value);
```

Comment:
nit: I think having `meanRate(Pair<Meter, ListenableFuture>)` and using specific names (e.g. `meanRate(tsLockLimited) + meanRate(tsLockDuplicatedLimited)`) should be easier to follow.

Hmm, it's a shame assertj doesn't have a nice way to write these constraints

Reply:
I like this a lot better, thanks!